resources/deliberate-ub.md

```diff
@@ -0,0 +1,41 @@
+When a crate explicitly acknowledges that what it does is UB, but prefers keeping that code over UB-free alternatives (or there are no UB-free alternatives), that is always a concerning sign.
+We should evaluate whether there truly is some use-case here that is not currently achievable in well-defined Rust, and work with crate authors on providing a UB-free solution.
+
+## Known cases of deliberate UB
+
+### Cases related to concurrency
+
+* crossbeam's `AtomicCell` implements an SeqLock, which [is well-known to not be compatible with the C++ memory model](http://www.hpl.hp.com/techreports/2012/HPL-2012-68.pdf).
+  Specifically the problem is [this non-atomic volatile read](https://github.com/crossbeam-rs/crossbeam/blob/5d07fe43540d7f21517a51813acd9332744e90cb/crossbeam-utils/src/atomic/atomic_cell.rs#L980) which can cause data races and hence UB.
+  This would be fine if we either (a) adopted LLVM's handling of memory races (then the problematic read would merely return `undef` instead of UB due to a data race), or (b) added [bytewise atomic memcpy](https://github.com/rust-lang/rfcs/pull/3301) and used that instead of the non-atomic volatile load.
+* crossbeam's `AtomicCell` also uses the standard library `Atomic*` types to do atomic reads and writes of [*any type* that has the right size](https://github.com/crossbeam-rs/crossbeam/blob/master/crossbeam-utils/src/atomic/atomic_cell.rs#L928-L932).
+  However, doing an `AtomicU32` atomic read on a `(u16, u8)` is unsound because the padding byte can be uninitialized.
+  (Also, pointer provenance is lost.)
+  To fix this we need to be able to perform atomic loads at type `MaybeUninit<u32>`.
+* Similar to the previous case, the `atomic-memcpy` crate uses the [standard `Atomic*` types to load potential padding or pointer bytes](https://github.com/taiki-e/atomic-memcpy/blob/b7b82bface0f24c5a9171f2d0626871c61302152/src/lib.rs#L323).
```

Comment:
Well, the atomic-memcpy's API is unsafe and passing a type containing uninitialized bytes violates its safety requirements.

Reply:
Ah, I hadn't realized that this is unsafe. I'll remove it then, thanks.

Reply:
Also, couldn't atomic-memcpy use https://github.com/taiki-e/atomic-maybe-uninit/ to avoid those safety requirements (potentially making it an option for use in `AtomicCell`)?

Reply:
> Also, couldn't atomic-memcpy use https://github.com/taiki-e/atomic-maybe-uninit/ to avoid those safety requirements (potentially making it an option for use in `AtomicCell`)?

https://github.com/taiki-e/atomic-memcpy/pull/7 is a PR for this, but it was pending because I thought the code generated could be more optimized on the atomic-maybe-uninit side (https://github.com/taiki-e/atomic-maybe-uninit/issues/1).